setuptools/dist.py

```diff
@@ -306,11 +306,18 @@ def __init__(self, attrs: Optional[MutableMapping] = None) -> None:
         dist_attrs = {k: v for k, v in attrs.items() if k not in metadata_only}
         _Distribution.__init__(self, dist_attrs)
 
-        # Private API (setuptools-use only, not restricted to Distribution)
+        # --- Private API (setuptools-use only, not restricted to Distribution) ---
+
         # Stores files that are referenced by the configuration and need to be in the
         # sdist (e.g. `version = file: VERSION.txt`)
         self._referenced_files: Set[str] = set()
 
+        # Stores a copy of the information that is consumed during configuration
+        # but that is still needed for commands to run (e.g. `dynamic` for `egg-info`)
+        self._config_info: Dict[str, Any] = {}
```

Comment:
This is a bit ugly... But when implementing this I could not find a better way of passing the information needed to where the checks should be performed:

1. The information about the original `pyproject.toml` was only known when parsing the config files. `dist.version` can be modified by other configuration files or `tool.setuptools.dynamic`... so we need to store the original config.